marshal/consumer.go

```diff
@@ -235,12 +235,47 @@ func (c *Consumer) defaultTopicPartitions() int {
 	return 0
 }
 
+// claimTerminated is called by a claim when they've terminated. This is used so we can
+// ensure topic claim semantics are adhered to. In topic claim mode this will be called
+// by every claim during a release.
+func (c *Consumer) claimTerminated(cl *claim, released bool) {
+	// For now, we don't care except in the topic claim mode
+	if !c.options.ClaimEntireTopic {
+		return
+	}
+
+	latestClaims, claimsErr := c.GetCurrentTopicClaims()
+	if claimsErr != nil {
+		// This only impacts update topic claims later
+		log.Errorf("claimTerminated: failed getting current claims: %s", claimsErr)
+	}
+	delete(latestClaims, cl.topic)
```

Comment:
I don't see how this is safe. We get a copy of our current claims, mutate our local copy, and only then get a lock to use that local copy to `updateTopicClaims`. If we gained a claim in the meantime before we get the lock it won't be reflected in `latestClaims`.